### PR TITLE
Adding more system tests for nifgen and updated function_addon

### LINKED ISSUE #484: Write system tests for NI-FGEN

### PR DESCRIPTION
**Pull request by injaleea**
[X] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).

~~[ ] I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.~~

~~[ ] I've added any unit tests that are applicable for this pull request~~

### What does this Pull Request accomplish?

Added system tests for fgen

- test_standard_waveform
- test_no_waveform_data
- test_frequency_list

And updated functions_addon.py to have proper enums for methods used in these system tests

### List issues fixed by this Pull Request below, if any.

Partial * Fix #484

### What testing has been done?

Verified generated signals using scope on dev machine, tox, system tests

### CODE REVIEW

File: src/nifgen/metadata/functions_addon.py [hunk 1]
```diff
@@ -35,6 +35,12 @@
 
 # Attach the given parameter to the given enum from enums.py
 functions_enums = {
+    'ConfigureOutputMode':               { 'parameters': { 1: { 'enum': 'OutputMode',                    }, }, },
```
Comment (marcoskirsch) [inline src/nifgen/metadata/functions_addon.py]: This conflicts with my work on #485.
I suggest you merge https://github.com/ni/nimi-python/tree/fgen-additional-metadata and then create system tests based on that.

Comment (marcoskirsch) [reply src/nifgen/metadata/functions_addon.py]: FYI there's a PR for this now
https://github.com/ni/nimi-python/pull/547

File: NI-FGEN notest for API.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
+FGEN notes
```
Comment (marcoskirsch) [inline NI-FGEN notest for API.txt]: remove this file

File: src/nifgen/system_tests/test_system_nifgen.py [hunk 3]
```diff
@@ -41,7 +41,7 @@ def test_method_get_self_cal_supported(session):
 
 def test_get_self_cal_last_date_and_time():
     try:
-        with nifgen.Session('', False, False, 'Simulate=1, DriverSetup=Model:5421;BoardType:PXI') as session:
+        with nifgen.Session('', False, 'Simulate=1, DriverSetup=Model:5421;BoardType:PXI') as session:
```
Comment (marcoskirsch) [inline src/nifgen/system_tests/test_system_nifgen.py]: * Document why this uses a 5421

File: src/nifgen/system_tests/test_system_nifgen.py [hunk 4]
```diff
@@ -50,3 +50,39 @@ def test_get_self_cal_last_date_and_time():
...
+    session.output_enabled = True
+    session.trigger_mode = nifgen.TriggerMode.NIFGEN_VAL_CONTINUOUS
+    session.trigger_source = nifgen.TriggerSource.NIFGEN_VAL_IMMEDIATE
+    with session.initiate():
+        assert session.func_amplitude == 2.0
+        assert session.func_waveform == nifgen.Waveform.SINE
+        assert session.is_done() is False
+
+
+def test_no_waveform_data(session):
```
Comment (marcoskirsch) [inline src/nifgen/system_tests/test_system_nifgen.py]: This test seems to be testing the driver, and not the Python bindings. Am I mistaken?
If so, I don't think we need it.

File: src/nifgen/system_tests/test_system_nifgen.py [hunk 5]
```diff
@@ -50,3 +50,39 @@ def test_get_self_cal_last_date_and_time():
...
+    session.clear_freq_list(-1)
+    duration_array = [0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01]
+    frequency_array = [1000, 100900, 200800, 300700, 400600, 500500, 600400, 700300, 800200, 900100]
+    waveform_handle = session.create_freq_list(nifgen.Waveform.SQUARE, frequency_array, duration_array)
+    session.configure_freq_list(waveform_handle, 2.0, 0, 0)
+    session.trigger_mode = nifgen.TriggerMode.NIFGEN_VAL_CONTINUOUS
+    session.output_enabled = True
+    with session.initiate():
+        assert session.func_waveform == nifgen.Waveform.SQUARE
+    session.output_enabled = True
```
Comment (marcoskirsch) [inline src/nifgen/system_tests/test_system_nifgen.py]: remove redundant calls

File: src/nifgen/metadata/functions_addon.py [hunk 6]
```diff
@@ -129,6 +129,5 @@
     'ConfigureDigitalEdgeStartTrigger':             { 'parameters': { 2: { 'default_value': 'StartTriggerDigitalEdgeEdge.RISING_EDGE', }, }, },
     'CreateAdvancedArbSequence':                    { 'parameters': { 4: { 'default_value': None, },
                                                                       5: { 'default_value': None, }, }, },
-    'WaitUntilDone':                                { 'parameters': { 1: { 'default_value': 10000, }, }, },
+    'WaitUntilDone': { 'parameters': { 1: { 'default_value': 10000, }, }, },
```
Comment (marcoskirsch) [inline src/nifgen/metadata/functions_addon.py]: please watch the spacing - not sure why this got messed up

File: src/nifgen/system_tests/test_system_nifgen.py [hunk 7]
```diff
@@ -41,7 +41,7 @@ def test_method_get_self_cal_supported(session):
 
 def test_get_self_cal_last_date_and_time():
     try:
-        with nifgen.Session('', False, 'Simulate=1, DriverSetup=Model:5421;BoardType:PXI') as session:
+        with nifgen.Session('', False, 'Simulate=1, DriverSetup=Model:5421;BoardType:PXI') as session:  # Simulated 5433 returns unrecoverable error when calling get_self_cal_last_date_and_time()
```
Comment (marcoskirsch) [inline src/nifgen/system_tests/test_system_nifgen.py]: this comment doesn't match the comment on line 48

File: src/nifgen/system_tests/test_system_nifgen.py [hunk 8]
```diff
@@ -50,3 +50,30 @@ def test_get_self_cal_last_date_and_time():
 
 def test_self_cal(session):
         session.self_cal()
+
+
+def test_standard_waveform(session):
+    session.output_mode = nifgen.OutputMode.NIFGEN_VAL_OUTPUT_FUNC
+    session.configure_standard_waveform(nifgen.Waveform.SINE, 2.0, 0.0, 0.0, 2000000)
```
Comment (marcoskirsch) [inline src/nifgen/system_tests/test_system_nifgen.py]: You should assert everything you set in this function below.
You should avoid using repeated values so you are sure the right parameter made it to the right thing.
What are you trying to accomplish with a start phase of 2 million degrees?
You should also test that some values have defaults.

File: src/nifgen/system_tests/test_system_nifgen.py [hunk 9]
```diff
@@ -50,3 +50,30 @@ def test_get_self_cal_last_date_and_time():
...
+    session.output_enabled = True
```
Comment (marcoskirsch) [inline src/nifgen/system_tests/test_system_nifgen.py]: I don't see the need to set attributes in lines 58, 59, 60 in this test.
I don't see the need to initiate.
Keep the tests as granular as possible. Do as little as you can.

File: src/nifgen/system_tests/test_system_nifgen.py [hunk 10]
```diff
@@ -50,3 +50,30 @@ def test_get_self_cal_last_date_and_time():
...
+    session.trigger_source = nifgen.TriggerSource.NIFGEN_VAL_IMMEDIATE
+    with session.initiate():
+        assert session.func_amplitude == 2.0
+        assert session.func_waveform == nifgen.Waveform.SINE
+        assert session.is_done() is False
+
+
+def test_frequency_list(session):
+    session.output_mode = nifgen.OutputMode.NIFGEN_VAL_OUTPUT_FREQ_LIST
+    session.clear_freq_list(-1)
```
Comment (marcoskirsch) [inline src/nifgen/system_tests/test_system_nifgen.py]: you should test this separately.

File: src/nifgen/system_tests/test_system_nifgen.py [hunk 11]
```diff
@@ -50,3 +50,30 @@ def test_get_self_cal_last_date_and_time():
...
+    duration_array = [0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01]
+    frequency_array = [1000, 100900, 200800, 300700, 400600, 500500, 600400, 700300, 800200, 900100]
+    waveform_handle = session.create_freq_list(nifgen.Waveform.SQUARE, frequency_array, duration_array)
+    session.configure_freq_list(waveform_handle, 2.0, 0, 0)
+    session.trigger_mode = nifgen.TriggerMode.NIFGEN_VAL_CONTINUOUS
+    session.output_enabled = True
+    with session.initiate():
+        assert session.func_waveform == nifgen.Waveform.SQUARE
+    with session.initiate():
```
Comment (marcoskirsch) [inline src/nifgen/system_tests/test_system_nifgen.py]: What are you trying to test by initiating twice?
Why are you even initiating?

Remember: The system tests should only test that the Python calls make it to the right C function. You should not be testing driver behavior. See https://github.com/ni/nimi-python/wiki/Testing-Strategy

File: src/nifgen/system_tests/test_system_nifgen.py (hunk 7 above)
Comment (injaleea) [reply src/nifgen/system_tests/test_system_nifgen.py]: Error codes are different for 5413/23/33 and other FGen
Only on 5413/23/33 calling some methods get_self_cal_last_date_and_time, get_ext_cal_last_date_and_time returns unrecoverable failure while other FGen devices returns operations not supported in simulated devices.

Comment (marcoskirsch) [reply src/nifgen/system_tests/test_system_nifgen.py]: That's ok, but the comment still doesn't match the comment on line 48.

**Pull request merged**